File: src/cache/ttl-manager.js
/**
 * Dynamic TTL Engine
 * Continuously recalibrates object time-to-live based on popularity, volatility, retrieval cost, and trend.
 */

import { DEFAULT_CONFIG } from '../core/types.js';

export class DynamicTTLManager {
  constructor(config = DEFAULT_CONFIG) {
    this.minTTL = config.minTTL || 15;
    this.maxTTL = config.maxTTL || 600;
    this.defaultTTL = config.defaultTTL || 90;
  }

  /**
   * Recalculates dynamic TTL for an entry
   */
  computeTTL(entry, evaluation) {
    const prevTTL = entry.currentTTL || this.defaultTTL;
    const { factors } = evaluation;

    // Base multipliers
    const popularityBonus = factors.popularity * 1.5;     // Up to +150%
    const retrievalCostBonus = factors.retrievalCost * 1.2; // Up to +120%
    const trendMultiplier = (factors.trend >= 0.5) ? (1 + (factors.trend - 0.5) * 1.0) : (0.5 + factors.trend);
    
    // Volatility penalty (if data mutates frequently, clamp TTL to avoid serving stale data)
    const volatility = entry.updateVolatility || 0.1;
    const volatilityPenalty = Math.max(0.3, 1.0 - (volatility * 0.8));

    // Dynamic TTL formula
    const rawTTL = this.defaultTTL * (1 + popularityBonus + retrievalCostBonus) * trendMultiplier * volatilityPenalty;
    const newTTL = Math.round(Math.max(this.minTTL, Math.min(this.maxTTL, rawTTL)));

    // Generate causal reason for change
    let reason = 'Steady-state baseline TTL';
    const delta = newTTL - prevTTL;

    if (Math.abs(delta) >= 10) {
      if (delta > 0) {
        if (factors.popularity > 0.6) {
          reason = `Extended (+${delta}s): Popularity surged & high retrieval cost (${entry.baseDbLatencyMs}ms).`;
        } else if (factors.trend > 0.6) {
          reason = `Extended (+${delta}s): Positive request trend detected.`;
        } else {
          reason = `Extended (+${delta}s): High utility reuse pattern.`;
        }
      } else {
        if (volatility > 0.5) {
          reason = `Shortened (${delta}s): High data volatility (${Math.round(volatility * 100)}% mutation rate).`;
        } else if (factors.trend < 0.4) {
          reason = `Shortened (${delta}s): Fading demand trend.`;
        } else {
          reason = `Shortened (${delta}s): Low reuse probability.`;
        }
      }
    }

    return {
      prevTTL,
      newTTL,
      ttlDelta: delta,
      reason,
      volatility
    };
  }
}

File: backend/cache/ttl_manager.py
"""
Dynamic TTL Recalibration Engine
Continuously adapts object TTL based on popularity, retrieval cost, trend velocity, and mutation volatility.
"""

from typing import Dict, Any
from backend.core.types import SystemConfig


class DynamicTTLManager:
    def __init__(
        self,
        min_ttl: int = SystemConfig.minTTL,
        max_ttl: int = SystemConfig.maxTTL,
        default_ttl: int = SystemConfig.defaultTTL
    ):
        self.min_ttl = min_ttl
        self.max_ttl = max_ttl
        self.default_ttl = default_ttl

    def compute_ttl(self, entry: Any, eval_result: Dict[str, Any]) -> Dict[str, Any]:
        factors = eval_result.get("factors", {})
        popularity = factors.get("popularity", 0.0)
        retrieval_cost = factors.get("retrievalCost", 0.0)
        trend = factors.get("trend", 0.5)
        
        volatility = getattr(entry, "update_volatility", 0.1)
        prev_ttl = getattr(entry, "current_ttl", self.default_ttl) or self.default_ttl

        # Dynamic TTL formula:
        # Base * (1 + 1.5*P + 1.2*C) * (0.5 + Trend) * (1 - 0.75*Volatility)
        base = float(self.default_ttl)
        popularity_boost = 1.0 + (1.5 * popularity) + (1.2 * retrieval_cost)
        trend_multiplier = 0.5 + trend
        volatility_penalty = max(0.2, 1.0 - (0.75 * volatility))

        raw_ttl = base * popularity_boost * trend_multiplier * volatility_penalty
        new_ttl = int(max(self.min_ttl, min(self.max_ttl, raw_ttl)))

        return {
            "prevTTL": prev_ttl,
            "newTTL": new_ttl,
            "delta": new_ttl - prev_ttl,
            "popularityBoost": round(popularity_boost, 2),
            "trendMultiplier": round(trend_multiplier, 2),
            "volatilityPenalty": round(volatility_penalty, 2)
        }
